File: Test/02_batch_annotation/test_batch_annotation.py
"""
CASSIA Test 02: Batch Annotation
================================
Tests the runCASSIA_batch function on all 6 cell clusters.

Usage:
    python test_batch_annotation.py
"""

import sys
import time
import os
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from fixtures import load_markers
from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    create_test_metadata
)

# Setup CASSIA imports
setup_cassia_imports()


def run_batch_annotation_test():
    """Run batch annotation test on all clusters."""
    print_test_header("02 - Batch Annotation")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Get LLM settings
    llm_config = config['llm']
    data_config = config['data']

    # Import CASSIA functions
    from tools_function import runCASSIA_batch

    # Use only 2 clusters for faster testing
    test_clusters = ['monocyte', 'plasma cell']

    # Load full marker data and filter to test clusters
    full_df = load_markers()
    marker_df = full_df[full_df['Broad.cell.type'].isin(test_clusters)].copy()

    print(f"\nTesting batch annotation for {len(test_clusters)} clusters:")
    for cluster in test_clusters:
        print(f"  - {cluster}")
    print(f"\nLoaded marker data: {marker_df.shape}")

    # Create results directory
    results_dir = create_results_dir("02_batch_annotation")
    output_name = str(results_dir / "batch_results")
    print(f"Results will be saved to: {results_dir}")

    # Run the test
    start_time = time.time()
    errors = []
    status = "error"

    try:
        print("\nRunning runCASSIA_batch...")
        runCASSIA_batch(
            marker=marker_df,
            output_name=output_name,
            n_genes=data_config.get('n_genes', 30),
            model=llm_config.get('model', 'google/gemini-2.5-flash'),
            temperature=llm_config.get('temperature', 0.3),
            tissue=data_config.get('tissue', 'large intestine'),
            species=data_config.get('species', 'human'),
            max_workers=llm_config.get('max_workers', 3),
            provider=llm_config.get('provider', 'openrouter'),
            validator_involvement=config.get('validator', {}).get('default', 'v1')
        )

        # Check output files
        full_csv = Path(f"{output_name}_full.csv")
        summary_csv = Path(f"{output_name}_summary.csv")

        if full_csv.exists():
            import pandas as pd
            results_df = pd.read_csv(full_csv)
            clusters_annotated = len(results_df)

            print(f"\nBatch Results:")
            print(f"  Clusters annotated: {clusters_annotated}/{len(test_clusters)}")
            print(f"  Output files created:")
            print(f"    - {full_csv.name}")
            if summary_csv.exists():
                print(f"    - {summary_csv.name}")

            if clusters_annotated == len(test_clusters):
                status = "passed"
            else:
                status = "failed"
                errors.append(f"Only {clusters_annotated}/{len(test_clusters)} clusters annotated")
        else:
            status = "failed"
            errors.append("Output file not created")

    except Exception as e:
        errors.append(str(e))
        status = "error"
        print(f"\nError: {e}")

    duration = time.time() - start_time

    # Save metadata
    metadata = create_test_metadata(
        test_name="batch_annotation",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=test_clusters,
        errors=errors
    )
    save_test_metadata(results_dir, metadata)

    # Print final result
    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    return success


if __name__ == "__main__":
    success = run_batch_annotation_test()
    sys.exit(0 if success else 1)

File: Test/02_batch_annotation/test_batch_annotation.R
# CASSIA Test 02: Batch Annotation (R)
# =====================================
# Tests the runCASSIA_batch function on all 6 cell clusters via R package.
#
# Usage:
#     Rscript test_batch_annotation.R

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  # Method 1: commandArgs (works with Rscript)
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  # Method 2: sys.frame (works with source())
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  # Fallback: current working directory
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "fixtures.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))

run_batch_annotation_test <- function() {
  print_test_header("02 - Batch Annotation (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Get settings
  llm_config <- config$llm
  data_config <- config$data

  # Get all clusters
  all_clusters <- get_all_clusters()

  # Create results directory (BEFORE any logging)
  results <- create_results_dir("02_batch_annotation", get_test_mode())
  start_logging(results$logs)

  log_msg("Testing batch annotation for", length(all_clusters), "clusters:")
  for (cluster in all_clusters) {
    log_msg("  -", cluster)
  }

  # Get full marker dataframe
  marker_df <- get_full_marker_dataframe()
  log_msg("Loaded marker data:", nrow(marker_df), "rows")

  output_name <- file.path(results$outputs, "batch_results")
  log_msg("Results will be saved to:", results$base)

  # Run the test
  start_time <- Sys.time()
  errors <- list()
  status <- "error"

  tryCatch({
    log_msg("Running runCASSIA_batch via R package...")

    # Call CASSIA R function
    CASSIA::runCASSIA_batch(
      marker = marker_df,
      output_name = output_name,
      n_genes = data_config$n_genes %||% 30,
      model = llm_config$model %||% "google/gemini-2.5-flash",
      temperature = llm_config$temperature %||% 0.3,
      tissue = data_config$tissue %||% "large intestine",
      species = data_config$species %||% "human",
      max_workers = llm_config$max_workers %||% 3,
      provider = llm_config$provider %||% "openrouter",
      validator_involvement = config$validator$default %||% "v1"
    )

    # Check output files
    full_csv <- paste0(output_name, "_full.csv")
    summary_csv <- paste0(output_name, "_summary.csv")

    if (file.exists(full_csv)) {
      results_df <- read.csv(full_csv)
      clusters_annotated <- nrow(results_df)

      log_msg("Batch Results:")
      log_msg("  Clusters annotated:", clusters_annotated, "/", length(all_clusters))
      log_msg("  Output files created:")
      log_msg("    -", basename(full_csv))
      if (file.exists(summary_csv)) {
        log_msg("    -", basename(summary_csv))
      }

      if (clusters_annotated == length(all_clusters)) {
        status <- "passed"
      } else {
        status <- "failed"
        errors <- list(paste("Only", clusters_annotated, "/", length(all_clusters), "clusters annotated"))
      }
    } else {
      status <- "failed"
      errors <- list("Output file not created")
    }

  }, error = function(e) {
    errors <<- list(e$message)
    status <<- "error"
    log_error(e)
  })

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Save metadata
  metadata <- create_test_metadata(
    test_name = "batch_annotation",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = as.list(all_clusters),
    errors = errors
  )
  save_test_metadata(results$outputs, metadata)

  # Print final result
  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))

  stop_logging()
  return(success)
}

# Run test
success <- run_batch_annotation_test()
quit(status = if (success) 0 else 1)
